language/move-prover/bytecode-to-boogie/src/dataflow_analysis.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Adapted from AbstractInterpreter for Bytecode, this module defines the data-flow analysis
+//! framework for stackless bytecode.
+
+use bytecode_verifier::absint::{AbstractDomain, JoinResult};
+use bytecode_verifier::control_flow_graph::{BlockId, ControlFlowGraph};
+use stackless_bytecode_generator::stackless_bytecode::StacklessBytecode;
+use std::collections::HashMap;
+
+pub struct BlockState<State: Clone> {
+    pub pre: State,
+    pub post: State,
+}
+
+pub type StateMap<State> = HashMap<BlockId, BlockState<State>>;
+
+/// Take a pre-state + instruction and mutate it to produce a post-state。
+pub trait TransferFunctions {
+    type State: AbstractDomain + Clone;
+
+    /// Execute local@instr found at index local@index in the current basic block from pre-state
+    /// local@pre.
+    /// Auxilary data from the analysis that is not part of the abstract state can be collected by
+    /// mutating local@self.
+    /// The last instruction index in the current block is local@last_index. Knowing this
+    /// information allows clients to detect the end of a basic block and special-case appropriately
+    fn execute(
+        &mut self,
+        pre: &mut Self::State,
```

Comment:
Is there a reason why we don't make this a function to State instead of letting it mutate the pre-state? If its a mutation model, I wonder why we need pre/post states.